### ISSUE TITLE
dask.distributed client hangs in VSCode

### ISSUE THREAD

**Issue by mgsnuno** [dask/dask#5525]
Assistance request for those using vscode (https://github.com/microsoft/vscode-python/issues/7845):

dask - 2.6.0
vscode - 1.39.2 (latest but version independent)
python extension - 2019.10.41019 (it's tied to this version but vscode team cannot find issue without diving into dask)

## Steps for bug
```python
from dask.distributed import Client
client = Client()
```
- start `Python Interactive` and paste this in input terminal and we get never ending stream of

```python
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
...
```
- restart notebook server (using interactive window) and run same code and it runs fine. 

If you run this code in python, ipython or jupyter notebook, no problems.

If anyone using vscode this is very easy to reproduce and if you can find a possible cause for the error I can open a new issue and hopefully point the team towards the right direction.

Thank you

**Comment by TomAugspurger** [comment]
`Client()` does quite a bit. To narrow this down, can you see if creating a LocalCluster also causes the issue?

```
from distributed import LocalCluster

LocalCluster()
```

Can you also try with LocalCluster(processes=False)?

**Comment by mgsnuno** [comment]
It also happens with `LocalCluster()`, with `LocalCluster(processes=False)` there are no problems. So it's tied to the processes.

**Comment by TomAugspurger** [comment]
It seems to work if you create the client in a `__main__` block

```python
if __name__ == '__main__':
    client = Client()

```

Can you confirm that?

**Comment by mgsnuno** [comment]
It also fails. 
If you select the code in a file and use shift+enter to run it, it fails always, with or without `__main__` block. 
If you paste the code in the interactive window terminal, it fails less but still fails sometimes. By failing I mean hanging and starting to print the warning messages.

**Comment by mrocklin** [comment]
One approach to resolve this problem would be to try to reproduce the error without Dask, and then raise that back on the vscode issue tracker and see if they can help.

You might try creating a process with the `multiprocessing` module, including using the `forkserver` context if necessary (assuming that you are not on Windows)

```python
import multiprocessing, time

process = multiprocessing.Process(target=time.sleep, args=(5,))
```


```python
import multiprocessing, time

context = multiprocessing.get_context("forkserver")
process = context.Process(target=time.sleep, args=(5,))
```

If both of those fail to reproduce the bug then we would try things of increasing complexity.  This would help to isolate the fundamental issue regardless.

**Comment by TomAugspurger** [comment]
There's an example in
https://docs.dask.org/en/latest/configuration.html#yaml-files

**Comment by mgsnuno** [comment]
```
import dask

dask.config.set(
    {
        "distributed.logging.distributed": "debug",
        "distributed.logging.bokeh": "debug",
        "distributed.logging.tornado": "debug",
    }
)
```

Nothing new, still the same warning messages

```
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
...
```

@mrocklin more complex things to run?

**Comment by TomAugspurger** [comment]
Perhaps a small tornado hello world application?

If that doesn't hang, then you may want to poke around where dask and
jupyter interact (this is only after
VSCode changed something jupyter-related, right?).

This sounds somewhat similar to
https://github.com/dask/distributed/pull/2343.

**Comment by mgsnuno** [comment]
Sorry for the delay.

Tornado hello world fails (using port `8786` because `8888` was in use):

```python
import tornado.ioloop
import tornado.web

class MainHandler(tornado.web.RequestHandler):
    def get(self):
        self.write("Hello, world")

def make_app():
    return tornado.web.Application([
        (r"/", MainHandler),
    ])

if __name__ == "__main__":
    app = make_app()
    app.listen(8786)
    tornado.ioloop.IOLoop.current().start()
```

with following error:

```
RuntimeError                              Traceback (most recent call last)
 in 
     14     app = make_app()
     15     app.listen(8786)
---> 16     tornado.ioloop.IOLoop.current().start()

~/miniconda3/envs/povmap/lib/python3.7/site-packages/tornado/platform/asyncio.py in start(self)
    146             self._setup_logging()
    147             asyncio.set_event_loop(self.asyncio_loop)
--> 148             self.asyncio_loop.run_forever()
    149         finally:
    150             asyncio.set_event_loop(old_loop)

~/miniconda3/envs/povmap/lib/python3.7/asyncio/base_events.py in run_forever(self)
    524         self._check_closed()
    525         if self.is_running():
--> 526             raise RuntimeError('This event loop is already running')
    527         if events._get_running_loop() is not None:
    528             raise RuntimeError(

RuntimeError: This event loop is already running
```

This runs fine in python and ipython, just fails in vscode.

It points towards my assumption that somehow vscode is spawning things twice somehow.

Can we narrow this example further? 

Regarding dask/distributed#2343 what can I change nowadays on that code section to try bypass the problem?

**Comment by TomAugspurger** [comment]
I'm not entirely sure...

https://github.com/dask/distributed/issues/3202 sounds somewhat related
though?

**Comment by jrbourbeau** [comment]
Things seemed to have stalled here. @mgsnuno is this still an issue for you, or is it okay to close?

**Comment by mgsnuno** [comment]
I stopped trying to find the issue, couldn't prioritize to work in it. Still, the problem is very much an issue.

After vscode-python introduced some changes (tied to `async` I would say), it has been broken since.

If I want to use `distributed.Client` in vscode python interactive window I have to restart the kernel a few times and copy+paste into the input terminal until the cell executes and I have a local cluster.

**Comment by gforsyth** [comment]
@mgsnuno -- the VSCode maintainers think this is an issue with `dask` (as noted in the upstream issue you opened with them) -- do you have a sense of which version of VSCode introduced the breaking changes?  One thing to check is whether an older version of VSCode can run a recent `LocalCluster`

**Comment by mgsnuno** [comment]
as I mentioned: 

> python extension - 2019.10.41019 (it's tied to this version but vscode team cannot find issue without diving into dask)

**Comment by devin-petersohn** [comment]
It there any update on this? We got a report that someone ran into the same issue trying to run Dask on VS code.

**Comment by rm4193** [comment]
> It there any update on this? We got a report that someone ran into the same issue trying to run Dask on VS code.

My issue is that whenever I try to import modin in VS code it is taking forever to import. Pandas seems to import faster. I tested importing modin on jupyter and in my terminal - it works perfectly fine. But for my application, I would need to use VS code. I also have all the Dask dependencies downloaded.  When running the debugging lines, it seems to not finish compiling. I am running on a Windows OS with 6 core CPU processor and NVIDIA GeForce GPU.

**Comment by mrocklin** [comment]
Modin creates a Dask Client at import time (this is not recommended
generally for libraries).

Dask Clients, by default, create processes.

VSCode's interpreter doesn't seem to work well in some cases when the
Python code within it creates processes.  I don't know the full context
about when this happens.

For other VSCode users, the typical advice is to create a client without
processes.

    client = Client(processes=False)

I'm not sure about the best way to do this with Modin though.

**Comment by devin-petersohn** [comment]
A client can be created before `import modin.pandas as pd` and Modin will use the existing `Client`. The `processes=False` keyword can lead to some out of memory errors getting thrown depending on the size of data. One thing you can try as a temporary workaround @rm4193 is this:

```python
from distributed import Client
client = Client(processes=False)
import modin.pandas as pd
```

We can see if this is working, but personally I have had issues with larger datasets. 

I do not want to hijack the thread to be about Modin, but maybe this will fix your issue @rm4193.

**Comment by mrocklin** [comment]
Or if anyone has the time to dive into what's happening between VSCode and Dask that would be great.  As I mentioned before in this thread, I think that the thing to do here is to try to replicate this problem with the multiprocessing module without Dask to see what the actual culprit is.

**Comment by harrisliuwk** [comment]
Just an update on identifying the problem.

platform: Windows 10 Version 1809
Dask: 2.20.0
VSCode: 1.47.2

It seems that the problem is compatibility between Dask and VSCode's Python Interactive Window.

When you do
```python
from distributed import Client
client = Client()
```
and
```python
from distributed import LocalCluster 
client = LocalCluster()
```
both work well with powershell and VSCode's Jupyter notebook environment. Note that the default for both `Client` and `LocalCluster` is `processes=True`.

But if you run the example codes above in Python Interactive Window, it just hangs forever, and there's no error message printed out.
Instead if you do
```python
from distributed import Client
client = Client(processed=False)
```
and
```python
from distributed import LocalCluster 
client = LocalCluster(processed=False)
```
in Interactive Window, the code will run as expected.

**Comment by mgsnuno** [comment]
@harrisliuwk I have experienced the same, it is tied to multiprocessing. 

I agree with @mrocklin to reproduce this using just the multiprocessing and asyncio modules and find the (timing) issues mentioned in https://github.com/microsoft/vscode-python/issues/7845#issuecomment-543990049.

I have been using dask-kubernetes in azure aks and stopped trying to find the issue.

**Comment by johnros-pagaya** [comment]
I am observing the same phenomenon.

**Comment by jrbourbeau** [comment]
https://github.com/dask/distributed/pull/4733 should have resolved the `AttributeError: 'NoneType' object has no attribute 'process'` issue (xref https://github.com/dask/distributed/issues/5386). I recommended folks try updating to the latest `dask` and `distributed` releases to see if that helps resolve the issue
